magento/magento2
# magento/magento2#37039: Not generating 301 redirects for product

fixed issue with generating 301 redirects for product after moving a category to a new parent

According to #37039 observer declared in  \Magento\CatalogUrlRewrite\Observer\CategoryProcessUrlRewriteMovingObserver.php has not proper order of actions for generating 301 redirects for products.

### Description (*)
- Fix for generating url rewrites for product before deleting category rewrites for children

### Related Pull Requests
no

### Fixed Issues (if relevant)
1. Fixes magento/magento2#37039

### Manual testing scenarios (*)
1 Go to Stores > Settings > Configuration > Catalog > Catalog > Search Engine Optimization
2. Make sure that configs: "Use Categories Path for Product URLs", "Generate "category/product" URL Rewrites" and "Create Permanent Redirect for URLs if URL Key Changed" set as "Yes"
3.Create 2 categories under the "Default" category,  on the same level and then create a new one as a sub-category of one that you just created.
4.Create a new product and assign it to the sub-category that you have created on step 3 of this guide
5.Open on a storefront your product, which should contins the path like `<Base URL>/<category-url-key>/<sub-category-url-key>/<product-url-key>.<Product URL Suffix>`
6.Go back to the admin and move with "drag and drop functionality" your sub-category to another category that you created on step 2.
7.Go to the storefront and reload your URL (from step 5) and it will redirect you to the new category(if you use my PR) or you wil get 404 error without this PR.

### Questions or commens
Additionaly I noticed that when indexer "catalog_category_product" configured as "update on save"  and if you act as I described in block "Manual testing scenarios", so in that case after moving sub-category for both created categories the indexer "catalogsearch_fulltext" not reindexed or even not invalidated. That's leading us to bad storefront look as the first category does not contain new products from recently moved sub-category and the second one stil contains those products in it. This issue is causing \Magento\Elasticsearch\Model\Indexer\Fulltext\Plugin\Category\Product\Action\Rows by this class on line 79 as there exists condition for scheduled indexer 
![image](https://user-images.githubusercontent.com/75758543/227541881-43ac49e6-9ca6-49b0-a5dd-327b7fd097d4.png)
Please let me know if it should be removed, so I will add a new PR for this as for now I'm confusing as it can increase save timeout Or should I invalidate related index?


### Contribution checklist (*)
